File: Coding-Interviews/19-正则表达式匹配.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;
class Solution
{
public:
    bool match(int a, int b, string s, string p) { return p[b] == '.' || s[a] == p[b]; }
    bool isMatch(string s, string p)
    {
        vector<vector<int>> dp(s.size() + 1, vector<int>(p.size() + 1, false));
        dp[0][0] = true;
        for (int i = 1; i < p.size() + 1; i++) {
            if (p[i - 1] == '*')
                dp[0][i] = dp[0][i - 2];
        }
        for (int i = 1; i < s.size() + 1; i++) {
            for (int j = 1; j < p.size() + 1; j++) {
                if (p[j - 1] == '*')
                    dp[i][j] =
                        dp[i][j - 2] || (dp[i - 1][j] && match(i - 1, j - 2, s, p)) || dp[i][j - 1];
                else if (p[j - 1] == '.') {
                    dp[i][j] = dp[i - 1][j - 1];
                }
                else
                    dp[i][j] = dp[i - 1][j - 1] && s[i - 1] == p[j - 1];
            }
        }
        return dp[s.size()][p.size()];
    }
};

File: Coding-Interviews/19-正则表达式匹配.py
class Solution:

    def isMatch(self, s: str, p: str) -> bool:
        # dp[i][j] s 的前i个和p的前j个是否匹配
        def match(s: str, p: str) -> bool:
            return s == p or p == '.'
        m = len(s)+1
        n = len(p)+1
        dp = [[False]*n for _ in range(m)]
        dp[0][0] = True
        for i in range(2, n):
            if i % 2 == 0 and p[i-1] == '*':
                dp[0][i] = dp[0][i-2]
        for i in range(1, m):
            for j in range(1, n):
                if p[j-1] == '.':
                    dp[i][j] = dp[i-1][j-1]
                elif p[j-1] == '*':
                    dp[i][j] = dp[i][j-1] or dp[i][j-2] or \
                        (dp[i-1][j] and match(s[i-1], p[j-2])) or \
                        (dp[i-1][j-1] and match(s[i-1], p[j-2]))
                else:
                    dp[i][j] = dp[i-1][j-1] and match(s[i-1], p[j-1])

        return dp[len(s)][len(p)]
